ksql-engine/src/main/java/io/confluent/ksql/planner/plan/StructuredDataSourceNode.java

```diff
@@ -255,10 +256,43 @@ private Integer findMatchingTimestampField(final String timestampFieldName) {
     return null;
   }
 
+  @SuppressWarnings("unchecked")
+  private KStream<?, GenericRow> createKStream(
+      final StreamsBuilder builder,
+      final TimestampExtractor timestampExtractor,
+      final Serde<GenericRow> genericRowSerde) {
+    final KsqlStream ksqlTable = (KsqlStream) getStructuredDataSource();
+
+    if (ksqlTable.isWindowed()) {
+      return stream(builder, timestampExtractor, genericRowSerde,
+          (Serde<Windowed<String>>)ksqlTable.getKeySerde(), windowedMapper);
+    }
+
+    return stream(builder, timestampExtractor, genericRowSerde,
+        (Serde<String>)ksqlTable.getKeySerde(), nonWindowedValueMapper);
+  }
+
+  private <K> KStream<K, GenericRow> stream(
+      final StreamsBuilder builder,
+      final TimestampExtractor timestampExtractor,
+      final Serde<GenericRow> genericRowSerde,
+      final Serde<K> keySerde,
+      final ValueMapperWithKey<K, GenericRow, GenericRow> mapper) {
+
+    final Consumed<K, GenericRow> consumed = Consumed
+        .with(keySerde, genericRowSerde)
+        .withTimestampExtractor(timestampExtractor);
+
+    return builder
+        .stream(getStructuredDataSource().getKsqlTopic().getKafkaTopicName(), consumed)
```

Comment:
nit: can we add a `getKafkaTopicName()` method to `getStructuredDataSource()` ? the chaining of `getXXXX` calls grates against me..

Reply:
Nope. This will undoubtedly break a lot of tests that mock stuff.  Deadlines approach I'm afraid I'm going to have to push back and say 'not right now'